File: backend/src/HIS.Core/Entities/EInvoice.cs
namespace HIS.Core.Entities;

/// <summary>
/// Hóa đơn điện tử (HĐĐT) đa NCC — VNPT / Viettel / MISA. Issue #24.
/// Bảng EInvoices (migration 135). KHÔNG FK cứng tới Receipts (tránh schema-drift).
/// </summary>
public class EInvoice : BaseEntity
{
    /// <summary>Liên kết phiếu thu (nullable — không FK cứng).</summary>
    public Guid? ReceiptId { get; set; }

    /// <summary>Nhà cung cấp: VNPT / Viettel / MISA.</summary>
    public string Provider { get; set; } = string.Empty;

    /// <summary>Số hóa đơn NCC cấp.</summary>
    public string? InvoiceNo { get; set; }

    /// <summary>Mã CQT (cơ quan thuế).</summary>
    public string? InvoiceCode { get; set; }

    /// <summary>Mẫu số hóa đơn.</summary>
    public string? TemplateCode { get; set; }

    /// <summary>Ký hiệu hóa đơn.</summary>
    public string? SerialNo { get; set; }

    /// <summary>0=draft, 1=issued, 2=signed, 3=cancelled, 4=failed.</summary>
    public int Status { get; set; }

    /// <summary>Tổng tiền hóa đơn (lấy từ Receipt.FinalAmount khi phát hành).</summary>
    public decimal TotalAmount { get; set; }

    /// <summary>Tiền thuế (nếu có).</summary>
    public decimal? TaxAmount { get; set; }

    /// <summary>Thời điểm phát hành.</summary>
    public DateTime? IssuedAt { get; set; }

    /// <summary>Raw response JSON từ NCC (giữ nguyên bản gốc).</summary>
    public string? PortalResponse { get; set; }

    /// <summary>Thông báo lỗi khi Status=4.</summary>
    public string? ErrorMessage { get; set; }
}

File: frontend/src/api/einvoice.ts
import apiClient from './client';

// ============================================================================
// HĐĐT — Hóa đơn điện tử đa NCC (VNPT / Viettel / MISA)
// Route BE: /api/einvoice
// MockMode mặc định true — sinh InvoiceNo giả, không gọi NCC thật.
// ============================================================================

export type EInvoiceStatus = 0 | 1 | 2 | 3 | 4;
// 0=draft, 1=issued, 2=signed, 3=cancelled, 4=failed

export interface EInvoiceDto {
  id: string;
  receiptId?: string;
  provider: string;          // "VNPT" | "Viettel" | "MISA"
  invoiceNo?: string;
  invoiceCode?: string;      // Mã CQT
  templateCode?: string;
  serialNo?: string;
  status: EInvoiceStatus;
  statusName: string;
  totalAmount: number;
  taxAmount?: number;
  issuedAt?: string;
  errorMessage?: string;
  createdAt: string;
}

export interface EInvoiceDetailDto extends EInvoiceDto {
  portalResponse?: string;   // Raw JSON từ NCC
}

export interface IssueEInvoiceDto {
  receiptId: string;
  /** NCC muốn dùng — nếu bỏ trống, lấy từ server config. */
  provider?: string;
}

export interface EInvoiceConfigDto {
  /** NCC đang kích hoạt: "VNPT" | "Viettel" | "MISA" */
  provider: string;
  /** true = sinh số giả, không gọi NCC */
  mockMode: boolean;
  enabled: boolean;
}

export interface EInvoiceListParams {
  receiptId?: string;
  status?: number;
  pageSize?: number;
}

// ── API object ───────────────────────────────────────────────────────────────

export const einvoice = {
  /** Lấy danh sách HĐĐT. */
  getList: async (params?: EInvoiceListParams): Promise<EInvoiceDto[]> => {
    const { data } = await apiClient.get<EInvoiceDto[]>('/einvoice', { params });
    return data;
  },

  /** Chi tiết 1 HĐĐT (kèm PortalResponse thô). */
  getDetail: async (id: string): Promise<EInvoiceDetailDto> => {
    const { data } = await apiClient.get<EInvoiceDetailDto>(`/einvoice/${id}`);
    return data;
  },

  /**
   * Phát hành HĐĐT từ phiếu thu.
   * MockMode=true (server default) → InvoiceNo giả "MOCK-…".
   * MockMode=false → gọi NCC thật (cần credential trong Cloud Run env).
   */
  issue: async (dto: IssueEInvoiceDto): Promise<EInvoiceDto> => {
    const { data } = await apiClient.post<EInvoiceDto>('/einvoice/issue', dto);
    return data;
  },

  /** Hủy HĐĐT. */
  cancel: async (id: string): Promise<EInvoiceDto> => {
    const { data } = await apiClient.post<EInvoiceDto>(`/einvoice/cancel/${id}`);
    return data;
  },

  /**
   * Đồng bộ trạng thái từ NCC.
   * MockMode=true → trả current status không gọi NCC.
   */
  syncStatus: async (id: string): Promise<EInvoiceDto> => {
    const { data } = await apiClient.post<EInvoiceDto>(`/einvoice/status/${id}`);
    return data;
  },

  /** Đọc cấu hình hiện tại (Provider, MockMode, Enabled). */
  getConfig: async (): Promise<EInvoiceConfigDto> => {
    const { data } = await apiClient.get<EInvoiceConfigDto>('/einvoice/config');
    return data;
  },

  /**
   * Lưu cấu hình.
   * Lưu ý: config không persist qua restart trừ khi set Cloud Run env.
   * Credential NCC cần set:
   *   EInvoice__Vnpt__BaseUrl, EInvoice__Vnpt__Account, EInvoice__Vnpt__Password,
   *   EInvoice__Vnpt__Serial, EInvoice__Vnpt__Pattern, EInvoice__Vnpt__TaxCode
   *   (tương tự cho Viettel__ và MISA__)
   */
  saveConfig: async (cfg: EInvoiceConfigDto): Promise<{ message: string }> => {
    const { data } = await apiClient.put<{ message: string }>('/einvoice/config', cfg);
    return data;
  },
};
